### PR TITLE
fix: 11304: Unit Test Failure: VirtualMapHashingTest#fullLeavesRehash fails intermittently

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary: fixed a thread race in `ConcurrentBlockingIterator` between `close()` and `hasNext()`. In some rare cases, a consumer thread calling `hasNext()` could return false, if a supplier thread supplied an element to the iterator and immediately after that closed the iterator.

A new unit test is provided.

Fixes: https://github.com/hashgraph/hedera-services/issues/11304
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

**Review by anthony-swirldslabs — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIteratorTest.java [hunk 1]
```diff
@@ -220,6 +221,28 @@ void multiThreadIteration() throws ExecutionException, InterruptedException {
...
+            } catch (final InterruptedException e) {
+                fail("Supplier interrupted");
+            }
+        });
+        supplier.start();
+        assertTrue(iterator.hasNext(), "Iterator must have more than zero elements");
+        assertEquals(1, iterator.next());
+        assertTrue(iterator.hasNext(), "Iterator must have more than one element");
+        assertEquals(2, iterator.next());
+        assertFalse(iterator.hasNext(), "Iterator must not have more than one elements");
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIteratorTest.java:242]: s/one/two/

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIteratorTest.java (hunk 1 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/test/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIteratorTest.java:242]: Nice catch. I will fix it

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java [hunk 2]
```diff
@@ -109,17 +109,15 @@ public boolean hasNext() {
...
         final long timeOutWhenMillisAre = System.currentTimeMillis() + waitMillis;
-        while ((next = buffer.poll()) == null) {
-            if (closed.get()) {
-                return false;
-            } else {
-                if (System.currentTimeMillis() > timeOutWhenMillisAre) {
-                    throw new RuntimeException(new TimeoutException("Timed out trying to read from buffer"));
-                }
+        boolean isOpen = !closed.get();
+        while (((next = buffer.poll()) == null) && isOpen) {
```
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: This is a good question. I am personally not sure this code is so performance critical as to justify the busy loop above. I'd expect this loop to help with local reconnect tests (where node transfers between teacher and learner are very fast), but may not be very helpful in real networks. Needs further investigation

**Review by OlegMazurov — APPROVED** [review]

**Review by imalygin — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (anthony-swirldslabs) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: Sounds good.

**Review by anthony-swirldslabs — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java (hunk 2 above)
Comment (jsync-swirlds) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: Given the time resolution for the wait is milliseconds, it _might_ be worth considering using the timeout version of [poll](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/LinkedBlockingQueue.html#poll(long,java.util.concurrent.TimeUnit)) with a 1 millisecond timeout.  That would still check the closed value every millisecond, but avoid the thousands of additional loops in between.

Also, could this be rewritten as follows (I think it makes sense, but you might have a reason for splitting it out)?
```
        while (((next = buffer.poll()) == null) && !closed.get()) {
```

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: There is a reason. If written like above, thread context switch may happen after the call to `poll()`, but before the call to `closed.get()`. This is exactly the problem addressed in this PR. With the way how it's currently in the PR, a call to `poll()` is guaranteed after `closed.get()` returns true.

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: Regarding the `poll()` calls, let me file a low-priority ticket about it. As I wrote above in response to @anthony-swirldslabs comments, the chances are we don't need a busy loop here at all

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/reconnect/ConcurrentBlockingIterator.java:113]: Filed https://github.com/hashgraph/hedera-services/issues/11691

**Pull request merged**